### PR TITLE
add no-commonjs-exports rule

### PR DESCRIPTION
**Pull request by unicop**
# Main notes
@neriyarden 

1. Checkout how I implemented the `fixer`, it might be hard to follow at first, but you will understand
2. Checkout how the `tests` `testRuler` is doing an amazing job in also checking the expected output including the fix!!!! 🤯

### CODE REVIEW

File: src/configs/js-with-ts.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 export = {
     plugins: ['mig2ts'],
     rules: {
-        'no-dynamic-import': 'warn'
+        'no-dynamic-import': ['warn', { esmodule: true }],
```
Comment (unicop) [inline src/configs/js-with-ts.ts:4]: actually this should be call no-dynamic-require, where `esmodule` also includes the dynamic import, maybe I should split these into 2 different rules

Comment (unicop) [reply src/configs/js-with-ts.ts:4]: eslint  recommendation is to keep rules the smallest possible,
eslint as a mechanism that is running eslint over the code 10 times, that way, when you add a `fixer` into your rule and your fix is causing another issue, eslint knows to fix it as well automatically @neriyarden

File: src/rules/no-commonjs-exports.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import { AST_NODE_TYPES, TSESTree } from "@typescript-eslint/types";
+import { createRule, ruleMessageTemplate } from "../util";
+
+export type Options = [];
+export type MessageIds = 'noCommonJSExports';
+
+const topLevelVariablesSet = new Set<string>()
```
Comment (neriyarden) [inline src/rules/no-commonjs-exports.ts:7]: is this to prevent redeclaration of variables declared in the top of a file?

**Review by neriyarden — APPROVED** [review]
💪🏻

File: src/configs/js-with-ts.ts (hunk 1 above)
Comment (unicop) [reply src/configs/js-with-ts.ts:4]: I'm opening a ticket

File: src/rules/no-commonjs-exports.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                }
+            },
+        }
+    }
+})
+
+const EXPORTS = "exports"
+
+/** Verifies the AssignmentStatement */
+function isAssignment(node: TSESTree.AssignmentExpression) {
```
[the remark this answers is not in the document]
Comment (unicop) [reply src/rules/no-commonjs-exports.ts:92]: yes once we will finish implementing the beta version, we will already have more knowledge and several rules to see if we are repeating ourselves

File: src/rules/no-commonjs-exports.ts (hunk 2 above)
Comment (unicop) [reply src/rules/no-commonjs-exports.ts:7]: but it does it not for this file, but for files ESLint is running linting on :) 

I'm using this helper to make sure I'm not fixing `exports.a = {}` into `export const a = {}` when `a` already defined in the `"module" scope of a file (module)

**Comment by unicop** [comment]
@neriyarden I'll take a second look, maybe I can do a good usage with the `scope` variable.

File: src/constants/index.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { ScopeType } from "@typescript-eslint/utils/dist/ts-eslint-scope";
+import type { CamelToSnakeCase, SnakeCaseToCamelCase } from "../shared/types";
+
+type FixedScopeType = { [key in Capitalize<SnakeCaseToCamelCase<ScopeType>>]: FixedKeys<key> }
```
Comment (unicop) [inline src/constants/index.ts:4]: I implemented here VERY advanced things of generics, you can take a look!

File: src/rules/no-commonjs-exports.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                linterMessage: `Calls to CommonJS "${EXPORTS}" should be replaced with ESM "export" syntax.`
+            })
+        },
+        schema: [],
+
+    },
+    create(context) {
+        return {
+            Program(_node) {
+                topLevelVariablesSet.clear()
```
Comment (unicop) [inline src/rules/no-commonjs-exports.ts:34]: clearing `Set` between `Program` (modules)

File: src/shared/types/generics.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+type Separator = ' ' | '-' | '_';
+type SnakeCaseSeparator = '-'
+// type SnakeToCamelCase<S extends string> =
```
Comment (unicop) [inline src/shared/types/generics.ts:3]: another version of snake to camel, lets keep it here for now

File: src/constants/index.ts (hunk 4 above)
Comment (neriyarden) [reply src/constants/index.ts:4]: 🤯🥴🤔

File: src/rules/no-commonjs-exports.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            },
+            MemberExpression(node) {
+                if (!isModuleScope(context.getScope())) {
+                    return;
+                }
+                if (doesMemberExpressionIsExports(node)) {
```
Comment (neriyarden) [inline src/rules/no-commonjs-exports.ts:40]: confusing naming. i think `isMemberExpressionAnExports` is clearer

File: src/rules/no-commonjs-exports.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        schema: [],
+
+    },
+    create(context) {
+        return {
+            Program(_node) {
+                topLevelVariablesSet.clear()
+            },
+            MemberExpression(node) {
+                if (!isModuleScope(context.getScope())) {
```
Comment (neriyarden) [inline src/rules/no-commonjs-exports.ts:37]: good idea to check at the top of every rule

File: src/rules/no-commonjs-exports.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                    return;
+                }
+                if (doesIdentifierIsExports(node)) {
+                    context.report({
+                        node,
+                        messageId: 'noCommonJSExports'
+                    })
+                }
+            },
+            // Collection all top-level variables to understand if automatic fix is possible
```
Comment (neriyarden) [inline src/rules/no-commonjs-exports.ts:76]:
```suggestion
            // Collect all top-level variables to check if automatic fix is possible
```

**Review by neriyarden — APPROVED** [review]
added a couple of comments

File: src/rules/no-commonjs-exports.ts (hunk 8 above)
Comment (unicop) [reply src/rules/no-commonjs-exports.ts:37]: that way I'm removing indentation and the code is cleaner
